File: packages/folio-core/domain/todo.go
package domain

import "time"

type TodoStatus string

const (
	TodoPending    TodoStatus = "pending"
	TodoInProgress TodoStatus = "in_progress"
	TodoDone       TodoStatus = "done"
	TodoBlocked    TodoStatus = "blocked"
	TodoCancelled  TodoStatus = "cancelled"
)

// IsTerminal reports whether no further work is expected on the todo.
func (s TodoStatus) IsTerminal() bool {
	return s == TodoDone || s == TodoCancelled
}

type Priority string

const (
	PriorityLow    Priority = "low"
	PriorityMedium Priority = "medium"
	PriorityHigh   Priority = "high"
)

type Todo struct {
	ID        TodoID
	ProjectID ProjectID
	// PlanID is optional: a todo can stand alone, outside any plan.
	PlanID PlanID
	// TicketID is optional: an empty value means the todo sits directly under
	// the project rather than under one of its tickets.
	TicketID TicketID
	Title    string
	Details  string
	Status   TodoStatus
	Priority Priority
	Tags     []string
	// Position orders todos within their plan; gaps are allowed.
	Position  int
	DependsOn []TodoID
	DueDate   *time.Time
	CreatedBy UserID
	CreatedAt time.Time
	UpdatedAt time.Time

	// Blocked is derived from DependsOn against the full set on read. It is
	// never persisted, and is distinct from the explicit TodoBlocked status
	// which the agent sets by hand.
	Blocked bool
}

// TodoFilter narrows a todo listing. Zero values mean "no restriction".
type TodoFilter struct {
	PlanID   PlanID
	TicketID TicketID
	Status   []TodoStatus
	Priority Priority
	Tags     []string
	Search   string
	Limit    int
	Offset   int
}

File: apps/web/src/core/domain/todo.ts
import type { Branded } from './branded'
import type { PlanId } from './plan'
import type { TicketId } from './ticket'
import type { ProjectId, UserId } from './project'

export type TodoId = Branded<string, 'TodoId'>

export const todoId = (value: string): TodoId => value as TodoId

export const TODO_STATUSES = ['pending', 'in_progress', 'done', 'blocked', 'cancelled'] as const

export type TodoStatus = (typeof TODO_STATUSES)[number]

export const PRIORITIES = ['low', 'medium', 'high'] as const

export type Priority = (typeof PRIORITIES)[number]

export type Todo = {
	readonly id: TodoId
	readonly projectId: ProjectId
	readonly ticketId: TicketId | undefined
	readonly planId: PlanId | undefined
	readonly title: string
	readonly details: string
	readonly status: TodoStatus
	readonly priority: Priority
	readonly tags: readonly string[]
	readonly position: number
	readonly dependsOn: readonly TodoId[]
	readonly dueDate: Date | undefined
	readonly createdBy: UserId | undefined
	readonly createdAt: Date
	readonly updatedAt: Date
}

export const isTerminal = (status: TodoStatus): boolean => status === 'done' || status === 'cancelled'
